ipa-core/src/report.rs

```diff
@@ -413,6 +434,186 @@ where
+    BK: SharedValue,
+    TV: SharedValue,
+    TS: SharedValue,
+    Replicated<BK>: Serializable,
+    Replicated<TV>: Serializable,
+    Replicated<TS>: Serializable,
+{
+    const ENCAP_KEY_MK_OFFSET: usize = 0;
+    const CIPHERTEXT_MK_OFFSET: usize = Self::ENCAP_KEY_MK_OFFSET + EncapsulationSize::USIZE;
+    // need to round up ciphertext length to nearest multiple of 16 since AES has block length 16
+    const ENCAP_KEY_BTT_OFFSET: usize = (Self::CIPHERTEXT_MK_OFFSET
+        + U16::USIZE
+        + ((<Replicated<BA64> as Serializable>::Size::USIZE) + 15))
+        & (!15);
+    const CIPHERTEXT_BTT_OFFSET: usize = Self::ENCAP_KEY_BTT_OFFSET + EncapsulationSize::USIZE;
+    // need to round up ciphertext length to nearest multiple of 16 since AES has block length 16
+
+    const EVENT_TYPE_OFFSET: usize = (Self::CIPHERTEXT_BTT_OFFSET
+        + U16::USIZE
+        + ((<Replicated<BK> as Serializable>::Size::USIZE
+            + <Replicated<TV> as Serializable>::Size::USIZE
+            + <Replicated<TS> as Serializable>::Size::USIZE)
+            + 15))
+        & (!15);
+    const SITE_DOMAIN_OFFSET: usize = Self::EVENT_TYPE_OFFSET + 4;
+
+    const TS_OFFSET: usize = 0;
+
+    const BK_OFFSET: usize = Self::TS_OFFSET + <Replicated<TS> as Serializable>::Size::USIZE;
+    const TV_OFFSET: usize = Self::BK_OFFSET + <Replicated<BK> as Serializable>::Size::USIZE;
+
+    pub fn encap_key_mk(&self) -> &[u8] {
+        &self.data[Self::ENCAP_KEY_MK_OFFSET..Self::CIPHERTEXT_MK_OFFSET]
+    }
+
+    pub fn mk_ciphertext(&self) -> &[u8] {
+        &self.data[Self::CIPHERTEXT_MK_OFFSET..Self::ENCAP_KEY_BTT_OFFSET]
+    }
+
+    pub fn encap_key_btt(&self) -> &[u8] {
+        &self.data[Self::ENCAP_KEY_BTT_OFFSET..Self::CIPHERTEXT_BTT_OFFSET]
+    }
+
+    pub fn btt_ciphertext(&self) -> &[u8] {
+        &self.data[Self::CIPHERTEXT_BTT_OFFSET..Self::EVENT_TYPE_OFFSET]
+    }
+
+    /// ## Panics
+    /// Only if a `Report` constructor failed to validate the contents properly, which would be a bug.
+    pub fn event_type(&self) -> EventType {
+        EventType::try_from(self.data[Self::EVENT_TYPE_OFFSET]).unwrap() // validated on construction
+    }
+
+    pub fn key_id(&self) -> KeyIdentifier {
+        self.data[Self::EVENT_TYPE_OFFSET + 1]
+    }
+
+    /// ## Panics
+    /// Never.
+    pub fn epoch(&self) -> Epoch {
+        u16::from_le_bytes(
+            self.data[Self::EVENT_TYPE_OFFSET + 2..Self::SITE_DOMAIN_OFFSET]
+                .try_into()
+                .unwrap(), // infallible slice-to-array conversion
+        )
+    }
+
+    /// ## Panics
+    /// Only if a `Report` constructor failed to validate the contents properly, which would be a bug.
+    pub fn site_domain(&self) -> &str {
+        std::str::from_utf8(&self.data[Self::SITE_DOMAIN_OFFSET..]).unwrap() // validated on construction
+    }
+
+    /// ## Errors
+    /// If the report contents are invalid.
+    pub fn from_bytes(bytes: B) -> Result<Self, InvalidReportError> {
```

Comment:
maybe we can validate the size as well?

Reply:
Following Andys function here. Don't we validate the size already implicitly due to the fact that we ask for EventType::try_from(bytes[Self::EVENT_TYPE_OFFSET])? so if EVENT_TYPE_OFFSET leads to an out of bounds, doesn't it already return an error? Further, site_domain does not seem to have a bound on the size.

Reply:
I have added an new error and check the length